File: c-src/torben.cpp
#include <cstdio>
#include <cassert>
#include <array>

// from https://github.com/facebook/ThreatExchange/blob/main/pdq/cpp/hashing/torben.cpp
float torben(float m[], int n) {
  int i, less, greater, equal;
  float min, max, guess, maxltguess, mingtguess;

  min = max = m[0];
  for (i = 1; i < n; i++) {
    if (m[i] < min)
      min = m[i];
    if (m[i] > max)
      max = m[i];
  }

  while (1) {
    guess = (min + max) / 2;
    less = 0;
    greater = 0;
    equal = 0;
    maxltguess = min;
    mingtguess = max;
    for (i = 0; i < n; i++) {
      if (m[i] < guess) {
        less++;
        if (m[i] > maxltguess)
          maxltguess = m[i];
      } else if (m[i] > guess) {
        greater++;
        if (m[i] < mingtguess)
          mingtguess = m[i];
      } else {
        equal++;
      }
    }
    if (less <= (n + 1) / 2 && greater <= (n + 1) / 2)
      break;
    else if (less > greater) {
      max = maxltguess;
    } else {
      min = mingtguess;
    }
  }
  if (less >= (n + 1) / 2) {
    return maxltguess;
  } else if (less + equal >= (n + 1) / 2) {
    return guess;
  } else {
    return mingtguess;
  }
}

int main()
{
    {
        auto f = std::to_array({10.0f, 20.0f});
        printf("%f\n", torben(f.data(), f.size()));
        assert(torben(f.data(), f.size()) == 10.0);
    }

    {
        auto f = std::to_array({10.0f, 20.0f, 30.0f});
        printf("%f\n", torben(f.data(), f.size()));
        assert(torben(f.data(), f.size()) == 20.0);
    }

    {
        auto f = std::to_array({10.0f, 20.0f, 30.0f, 40.0f});
        printf("%f\n", torben(f.data(), f.size()));
        assert(torben(f.data(), f.size()) == 20.0);
    }

    {
        auto f = std::to_array({1.0f, 5.0f, 2.0f, 4.0f, 3.0f, 1.0f});
        printf("%f\n", torben(f.data(), f.size()));
        assert(torben(f.data(), f.size()) == 2.0);
    }
}

File: src/pdq/torben.zig
const std = @import("std");

/// The following code is public domain.
/// Algorithm by Torben Mogensen, implementation by N. Devillard.
pub fn torben(m: []const f32) f32 {
    var less: i32 = undefined;
    var greater: i32 = undefined;
    var equal: i32 = undefined;
    var min: f32 = undefined;
    var max: f32 = undefined;
    var guess: f32 = undefined;
    var maxltguess: f32 = undefined;
    var mingtguess: f32 = undefined;
    const n = m.len;

    max = m[0];
    min = max;
    for (m[1..]) |val| {
        if (val < min) {
            min = val;
        }
        if (val > max) {
            max = val;
        }
    }

    while (true) {
        guess = (min + max) / 2;
        less = 0;
        greater = 0;
        equal = 0;
        maxltguess = min;
        mingtguess = max;
        for (m[0..]) |val| {
            if (val < guess) {
                less += 1;
                if (val > maxltguess) {
                    maxltguess = val;
                }
            } else if (val > guess) {
                greater += 1;
                if (val < mingtguess) {
                    mingtguess = val;
                }
            } else {
                equal += 1;
            }
        }

        if ((less <= (n + 1) / 2) and (greater <= (n + 1) / 2)) {
            break;
        } else if (less > greater) {
            max = maxltguess;
        } else {
            min = mingtguess;
        }
    }

    if (less >= ((n + 1) / 2)) {
        return maxltguess;
    } else if (less + equal >= ((n + 1) / 2)) {
        return guess;
    } else {
        return mingtguess;
    }
}

test "torben: test 1 element" {
    const arr = [_]f32{10};
    try std.testing.expectEqual(10, torben(&arr));
}

test "torben: test 2 elements" {
    const arr = [_]f32{ 10, 20 };
    try std.testing.expectEqual(10, torben(&arr));
}

test "torben: test 3 elements" {
    const arr = [_]f32{ 10, 20, 30 };
    try std.testing.expectEqual(20, torben(&arr));
}

test "torben: test 4 elements" {
    const arr = [_]f32{ 10, 20, 30, 40 };
    try std.testing.expectEqual(20, torben(&arr));
}

test "torben: test 5 elements out of order" {
    const arr = [_]f32{ 1, 5, 2, 4, 3, 1 };
    try std.testing.expectEqual(2, torben(&arr));
}
